src/aws-resource-mappings.ts

```diff
@@ -92,75 +88,92 @@ export function mapToAwsResource(
                 throw new Error('QueuePolicy has an invalid value for `queues` property');
             }
 
-            return (props.queues || []).flatMap((q: string) => {
-                return new aws.sqs.QueuePolicy(logicalId, {
-                    policy: rawProps.PolicyDocument,
-                    queueUrl: q,
-                });
+            const queues: string[] = props.queues ?? [];
+            return queues.flatMap((q: string, i: number) => {
+                const id = i === 0 ? logicalId : `${logicalId}-policy-${i}`;
+                return {
+                    logicalId: id,
+                    resource: new aws.sqs.QueuePolicy(id, {
+                        policy: rawProps.PolicyDocument,
+                        queueUrl: q,
+                    }),
+                };
             });
         }
 
         // SNS
         // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sns-topicpolicy.html
-        case 'AWS::SNS::TopicPolicy':
+        case 'AWS::SNS::TopicPolicy': {
             if (!Array.isArray(props.topics)) {
                 throw new Error('TopicPolicy has an invalid value for `topics` property');
             }
 
-            return (props.topics || []).flatMap((arn: string) => {
-                return new aws.sns.TopicPolicy(logicalId, {
-                    policy: rawProps.PolicyDocument,
-                    arn,
-                });
+            const topics: string[] = props.topics ?? [];
+            return topics.flatMap((arn: string, i: number) => {
+                const id = i === 0 ? logicalId : `${logicalId}-policy-${i}`;
+                return {
+                    logicalId: id,
+                    resource: new aws.sns.TopicPolicy(id, {
+                        policy: rawProps.PolicyDocument,
+                        arn,
+                    }),
+                };
             });
+        }
 
         // IAM
         case 'AWS::IAM::Policy': {
-            const resources: ResourceMapping[] = [];
+            const resources: ResourceAttributeMappingArray = [];
             const policy = new aws.iam.Policy(
                 logicalId,
                 {
                     policy: rawProps.PolicyDocument,
                 },
                 options,
             );
-            resources.push(policy);
+            resources.push({
+                resource: policy,
+                logicalId,
+            });
 
             for (let i = 0; i < (props.groups || []).length; i++) {
-                resources.push(
-                    new aws.iam.GroupPolicyAttachment(
+                resources.push({
```

Comment:
Maybe that's just personal preference, but I feel like this would be more readable if we replaced the for loop with `props.groups?.map((group, idx) => {...}) ?? []`.

But as I said, just a nit